# Analyzing the Chinook Database with SQLITE


In [2]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db
    

'Connected: None@chinook.db'

## Exlopring the Data


Get all the tables and views in the database


In [8]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ('table', 'view');
    

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

In [27]:
%%sql
SELECT distinct *
FROM customer c, invoice i
WHERE c.customer_id = i.customer_id 
order by customer_id;


Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 16, 1, '2017-01-26 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 8.91),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 77, 1, '2017-05-28 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 5.9399999999999995),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 149, 1, '2017-11-30 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 8.91),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 153, 1, '2017-12-14 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 13.86),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 182, 1, '2018-02-07 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 5.9399999999999995),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 184, 1, '2018-02-09 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 0.99),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 223, 1, '2018-05-04 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 17.82),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 270, 1, '2018-08-27 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 10.89),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 296, 1, '2018-12-10 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 12.87),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 442, 1, '2019-11-05 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 3.96),
 (1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 464, 1, '2019-12-31 00:00:00', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', 3.96),

In [233]:
%%sql
SELECT 
    c.customer_id,
    count(c.customer_id) as total_per_customer
FROM customer c, invoice i
WHERE c.customer_id = i.customer_id 
GROUP BY c.customer_id
order by total_per_customer;

Done.


[(29, 4),
 (56, 5),
 (8, 7),
 (16, 8),
 (18, 8),
 (24, 8),
 (32, 8),
 (52, 8),
 (59, 8),
 (3, 9),
 (4, 9),
 (7, 9),
 (15, 9),
 (19, 9),
 (27, 9),
 (38, 9),
 (39, 9),
 (40, 9),
 (41, 9),
 (47, 9),
 (54, 9),
 (9, 10),
 (11, 10),
 (14, 10),
 (23, 10),
 (25, 10),
 (28, 10),
 (37, 10),
 (45, 10),
 (48, 10),
 (49, 10),
 (51, 10),
 (55, 10),
 (2, 11),
 (12, 11),
 (21, 11),
 (31, 11),
 (36, 11),
 (42, 11),
 (44, 11),
 (50, 11),
 (53, 11),
 (6, 12),
 (10, 12),
 (17, 12),
 (20, 12),
 (22, 12),
 (26, 12),
 (33, 12),
 (43, 12),
 (1, 13),
 (30, 13),
 (34, 13),
 (46, 13),
 (57, 13),
 (58, 13),
 (13, 15),
 (35, 16),
 (5, 18)]

In [63]:
%%sql
with customer_info as (
SELECT 
    c.customer_id,
    count(c.customer_id) as total_per_customer
FROM customer c, invoice i
WHERE c.customer_id = i.customer_id 
GROUP BY c.customer_id
order by c.customer_id
)
SELECT 
    min(total_per_customer),
    max(total_per_customer),
    avg(total_per_customer)
FROM customer_info;

Done.


[(4, 18, 10.40677966101695)]

There is a total of 59 customers who made a purchase, some of them made more than one purchase therefore multiple invoices. The average number of purchases is 10.4, which ranges between 4 and 18 purchases in total.

## Finding Best Genres for USA Market 

A new records comnpany wants to get in the American market. For advertising purposes, they want to know which are the top 3 best selling jenres in the USA.

In [129]:
%%sql

WITH usa_sold AS
   (
    SELECT il.* FROM invoice_line il
    INNER JOIN invoice i on il.invoice_id = i.invoice_id
    INNER JOIN customer c on i.customer_id = c.customer_id
    WHERE c.country = "USA"
   )
SELECT 
    g.name,
    count(invoice_id) as total_invoices,
    round(
        (cast(count(susa.invoice_id) AS FLOAT) / (
        SELECT COUNT(*) 
        FROM usa_sold ) * 100)
        , 2)
     percentages_sold
FROM usa_sold susa
INNER JOIN track t on t.track_id = susa.track_id
INNER JOIN genre g on g.genre_id = t.genre_id
GROUP BY g.name
ORDER BY total_invoices desc
LIMIT 3;

Done.


[('Rock', 561, 53.38),
 ('Alternative & Punk', 130, 12.37),
 ('Metal', 124, 11.8)]

Based on the findings above, the top 3 best sellers genres in the USA are Rock, Alternative & Punk and Metal.
The Rock genre is in the lead with 53% of all sales.


## Analyze Employees Sales Performance

In [215]:
%%sql

WITH employee_info AS (
SELECT
    e.employee_id,
    e.first_name || ' ' || e.last_name AS employee_name,
    e.title,
    e.reports_to,
    e.hire_date,
    il.invoice_id,
    il.unit_price,
    il.quantity,
    i.total
FROM employee e
LEFT JOIN customer c ON c.support_rep_id = e.employee_id
LEFT JOIN invoice i ON i.customer_id = c.customer_id
LEFT JOIN invoice_line il ON il.invoice_line_id = i.invoice_id
    )

SELECT 
    employee_name,
    round(sum(total), 2) as total_sales
FROM employee_info 
GROUP BY employee_id 
ORDER BY total_sales DESC
LIMIT 3;


Done.


[('Jane Peacock', 1731.51),
 ('Margaret Park', 1584.0),
 ('Steve Johnson', 1393.92)]

In [214]:
%%sql
WITH employee_info AS (
SELECT
    e.employee_id,
    e.first_name || ' ' || e.last_name AS employee_name,
    e.title,
    e.reports_to,
    e.hire_date,
    il.invoice_id,
    il.unit_price,
    il.quantity,
    i.total
FROM employee e
LEFT JOIN customer c ON c.support_rep_id = e.employee_id
LEFT JOIN invoice i ON i.customer_id = c.customer_id
LEFT JOIN invoice_line il ON il.invoice_line_id = i.invoice_id
    )

SELECT 
    employee_name,
    hire_date
FROM employee_info 
GROUP BY employee_id 
ORDER BY hire_date asc;


Done.


[('Nancy Edwards', '2016-05-01 00:00:00'),
 ('Andrew Adams', '2016-08-14 00:00:00'),
 ('Michael Mitchell', '2016-10-17 00:00:00'),
 ('Robert King', '2017-01-02 00:00:00'),
 ('Laura Callahan', '2017-03-04 00:00:00'),
 ('Jane Peacock', '2017-04-01 00:00:00'),
 ('Margaret Park', '2017-05-03 00:00:00'),
 ('Steve Johnson', '2017-10-17 00:00:00')]

In [231]:
%%sql
SELECT 
    e.first_name || ' ' || e.last_name AS 'Employee',
    m.first_name || ' ' || m.last_name AS 'Manager'
from employee e, employee m
WHERE e.reports_to = m.employee_id;




Done.


[('Nancy Edwards', 'Andrew Adams'),
 ('Jane Peacock', 'Nancy Edwards'),
 ('Margaret Park', 'Nancy Edwards'),
 ('Steve Johnson', 'Nancy Edwards'),
 ('Michael Mitchell', 'Andrew Adams'),
 ('Robert King', 'Michael Mitchell'),
 ('Laura Callahan', 'Michael Mitchell')]

Out of the 8 employees, Jane has the most sales amount (in US$).
Five employees have never made a sale, and could be new and unexperienced.
Jane is indeed the most experienced seller.
Andrew Adams reports to no one, and he is probably the CEO of the company, and does not selling records.


## Analyze Sales by Country

Next, we will analyize the sales by countries.
All countries with a single customer will be grouped into a united group.

In [337]:
#count number of customers per country
%%sql
 WITH countries AS (
    SELECT 
        c.country,
        count(c.customer_id) total_customers
    FROM customer c
    GROUP BY c.country
    ORDER BY total_customers desc
 )
SELECT *
FROM countries
     

Done.


[('USA', 13),
 ('Canada', 8),
 ('Brazil', 5),
 ('France', 5),
 ('Germany', 4),
 ('United Kingdom', 3),
 ('Czech Republic', 2),
 ('India', 2),
 ('Portugal', 2),
 ('Argentina', 1),
 ('Australia', 1),
 ('Austria', 1),
 ('Belgium', 1),
 ('Chile', 1),
 ('Denmark', 1),
 ('Finland', 1),
 ('Hungary', 1),
 ('Ireland', 1),
 ('Italy', 1),
 ('Netherlands', 1),
 ('Norway', 1),
 ('Poland', 1),
 ('Spain', 1),
 ('Sweden', 1)]

Group countries with one customer into 'Other', and find fore each country:
1 - The total number of customers 
2 - The total sales value 
3 - The average order value 
4 - The average order value 

In [70]:
%%sql
WITH countries_grouped as (
    SELECT 
        CASE WHEN COUNTRY IN(
                SELECT country FROM (
                SELECT country, count(*) as total
                FROM customer
                GROUP BY country )
                where total = 1) then 'Other'
        ELSE c.country
        END AS countries,
        customer_id
    FROM customer c 
)

SELECT 
    countries,
    total_customers,
    total_sales,
    avg_order_value,
    customer_lifetime_value
    
FROM (
    SELECT
        countries, 
        count(distinct cg.customer_id) total_customers,
        round(sum(i.total) ,2) total_sales,
        round(sum(il.unit_price) / count(distinct i.invoice_id) ,2) avg_order_value,
        round(sum(il.unit_price) / count(distinct cg.customer_id) ,2) customer_lifetime_value,
            CASE 
                WHEN countries = 'Other' THEN 1
                ELSE 0
                END AS sort_other
    FROM countries_grouped cg
    JOIN invoice i ON i.customer_id = cg.customer_id
    JOIN invoice_line il ON il.invoice_id = i.invoice_id
    GROUP BY countries
    
)

ORDER BY sort_other, total_sales desc




    
   

Done.


[('USA', 13, 10405.89, 7.94, 80.04),
 ('Canada', 8, 5489.55, 7.05, 66.95),
 ('Brazil', 5, 4059.0, 7.01, 85.54),
 ('France', 5, 3972.87, 7.78, 77.81),
 ('Germany', 4, 3441.24, 8.16, 83.66),
 ('Czech Republic', 2, 3183.84, 9.11, 136.62),
 ('United Kingdom', 3, 2498.76, 8.77, 81.84),
 ('India', 2, 1887.93, 8.72, 91.58),
 ('Portugal', 2, 1755.27, 6.38, 92.57),
 ('Other', 15, 10808.82, 7.45, 73.0)]

In [72]:
%%sql
    SELECT
    round(sum(unit_price) / count(distinct cg.customer_id) ,1) avg_customer_lifetime_value
FROM customer cg
    JOIN invoice i ON i.customer_id = cg.customer_id
    JOIN invoice_line il ON il.invoice_id = i.invoice_id


Done.


[(79.8,)]

Based on the data, the countries with the highest customer life time value are Czech Republic, India and the UK.
However, this data may not be representative due to low customers per country (2 customers in the Czech republic, 3 in the UK and 2 in India). All three countries have a higher than average customer life time value.
The average overall customer life time value is 79.8 dollars.